snuba/web/split.py

```diff
@@ -1,34 +1,153 @@
+    OPERATOR_TO_FUNCTION,
+    BooleanFunctions,
+    binary_condition,
+    in_condition,
+    is_binary_condition,
+)
+from snuba.query.dsl import literals_tuple
+from snuba.query.expressions import Column, Expression, FunctionCall, Literal
 from snuba.request.request_settings import RequestSettings
 from snuba.util import is_condition
+from snuba.utils.metrics.backends.wrapper import MetricsWrapper
 from snuba.web import QueryResult
 
+metrics = MetricsWrapper(environment.metrics, "query.splitter")
+
 # Every time we find zero results for a given step, expand the search window by
 # this factor. Based on the assumption that the initial window is 2 hours, the
 # worst case (there are 0 results in the database) would have us making 4
 # queries before hitting the 90d limit (2+20+200+2000 hours == 92 days).
 STEP_GROWTH = 10
 
 
-def _identify_condition(condition: Any, field: str, operand: str) -> bool:
-    return is_condition(condition) and condition[0] == field and condition[1] == operand
+def _identify_condition(condition: Any, field: str, operator: str) -> bool:
+    return (
+        is_condition(condition) and condition[0] == field and condition[1] == operator
+    )
+
+
+def _get_time_range(
+    query: Query, timestamp_field: str
+) -> Tuple[Optional[Literal], Optional[Literal]]:
+    """
+    Finds the time range for this query. Which means, finds the timestamp condition
+    with the highest datetime literal and the timestamp condition with the smallest
+    and returns the interval in the form of a tuple of Literals. It only looks into
+    first level AND conditions.
+
+    TODO: Consider making this part of the AST manipulation api if there are more use
+    cases. It would require managing a few more corner cases for being part of the api.
+    """
+
+    def get_first_level_conditions(exp: Expression) -> Sequence[Expression]:
+        """
+        Traverse the tree as long as it runs into AND conditions and return all the
+        top level conditions.
+        """
+        if isinstance(exp, FunctionCall) and is_binary_condition(
+            exp, BooleanFunctions.AND
+        ):
+            return [
+                *get_first_level_conditions(exp.parameters[0]),
+                *get_first_level_conditions(exp.parameters[1]),
+            ]
+        else:
+            return [exp]
+
+    condition_clause = query.get_condition_from_ast()
+    if not condition_clause:
+        return (None, None)
+
+    max_lower_bound = None
+    mix_upper_bound = None
+    for c in get_first_level_conditions(condition_clause):
+        if (
+            isinstance(c, FunctionCall)
+            and c.function_name
+            in (OPERATOR_TO_FUNCTION[">="], OPERATOR_TO_FUNCTION["<"])
+            and isinstance(c.parameters[0], Column)
+            and c.parameters[0].column_name == timestamp_field
+            and isinstance(c.parameters[1], Literal)
+            and isinstance(c.parameters[1].value, datetime)
```

Comment:
To me this is a use case where we could introduce a regex like api to match structures in an expression:

```
match(expression, function(operator(">="), (column("name"), param("timestamp")))
```

Reply:
Here it is: https://github.com/getsentry/snuba/pull/942